New CMakeDeps issues

Hi @memsharded, I've tried the new CMakeDeps generator, and ran into some problems: 

- Packages that use `cmake_find_mode=none` with a `fooConfig.cmake` file in the `cpp_info.builddirs` path (https://docs.conan.io/2/reference/tools/cmake/cmakedeps.html#disable-cmakedeps-for-installed-cmake-configuration-files) are no longer discovered. This path will need to be added to `CMAKE_FIND_ROOT_PATH` by `conan_cmakedeps_paths.cmake`. Adding them to `CMAKE_PREFIX_PATH` alone would not be sufficient in the case of cross-compilation, unless the `~/.conan2/p/b/fooabc/p` is already in the `CMAKE_FIND_ROOT_PATH`.
- There appears to be something wrong with the capitalization of the generated CMake targets. For example, when building a package `foo` that depends on OpenBLAS, the generated target `foo::foo` by Conan depends on `openblas::openblas`, whereas the actual generated OpenBLAS target is `OpenBLAS::OpenBLAS`.

If useful, I could open new issues with more information.

_Originally posted by @tttapa in https://github.com/conan-io/conan/issues/16324#issuecomment-2479341117_

Thanks!

Here's an example of the second issue: https://github.com/tttapa/conan-recipes/blob/11765d09410927dc001bb08de3f1a850e4b3e756/recipes/coinmumps/all/conanfile.py

```sh
git clone https://github.com/tttapa/conan-recipes
conan remote add tttapa-conan-recipes "$PWD/conan-recipes" --force
cd conan-recipes/recipes/ipopt/all
conan create . --version=3.14.16 --build=missing    # works
conan create . --version=3.14.16 --build=missing -c tools.cmake.cmakedeps:new=will_break_next    # does not work
```

<details><summary>Full output</summary>

```
======== Exporting recipe to the cache ========
ipopt/3.14.16: Exporting package recipe: /tmp/conan-recipes/recipes/ipopt/all/conanfile.py
ipopt/3.14.16: exports: File 'conandata.yml' found. Exporting it...
ipopt/3.14.16: Copied 1 '.yml' file: conandata.yml
ipopt/3.14.16: Copied 1 '.py' file: conanfile.py
ipopt/3.14.16: Exported to cache folder: /home/pieter/.conan2/p/ipopt9a746552a1ee1/e
ipopt/3.14.16: Exported: ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c (2024-11-15 18:14:35 UTC)

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux
[conf]
tools.cmake.cmakedeps:new=will_break_next

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux


======== Computing dependency graph ========
Graph root
    cli
Requirements
    coinmumps/3.0.7#53df8e3d1b90efec14bba86a1bf391c2 - Cache
    ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c - Cache
    openblas/0.3.27#c16c3dbdf45e4a4ff56f8ada8631a2f3 - Cache

======== Computing necessary packages ========
openblas/0.3.27: Main binary package 'bc6c7a69ae72a75e92e2dcbdc9e3560daef36419' missing
openblas/0.3.27: Checking 11 compatible configurations
openblas/0.3.27: Found compatible package 'eb6292a0e957646b54c2bfd72a2a6c8d0b7a1aac': compiler.cppstd=17
ipopt/3.14.16: Forced build from source
Requirements
    coinmumps/3.0.7#53df8e3d1b90efec14bba86a1bf391c2:da8cda98ce5c347f86e367506d27090df7150d5e#17436017e49b2b4c6eb2d8c6b264cc14 - Cache
    ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c:aee99e0e9ca0db3bc3f87a71d61629d23c32e9b4 - Build
    openblas/0.3.27#c16c3dbdf45e4a4ff56f8ada8631a2f3:eb6292a0e957646b54c2bfd72a2a6c8d0b7a1aac#6231bd84d94ddb4211bfaa4d6287e04c - Cache

======== Installing packages ========
openblas/0.3.27: Already installed! (1 of 3)
coinmumps/3.0.7: Already installed! (2 of 3)
ipopt/3.14.16: Calling source() in /home/pieter/.conan2/p/ipopt9a746552a1ee1/s
ipopt/3.14.16: Unzipping 3.14.16.tar.gz to /home/pieter/.conan2/p/ipopt9a746552a1ee1/s

-------- Installing package ipopt/3.14.16 (3 of 3) --------
ipopt/3.14.16: Building from source
ipopt/3.14.16: Package ipopt/3.14.16:aee99e0e9ca0db3bc3f87a71d61629d23c32e9b4
ipopt/3.14.16: Copying sources to build folder
ipopt/3.14.16: Building your package in /home/pieter/.conan2/p/b/ipopte17a619841258/b
ipopt/3.14.16: Calling generate()
ipopt/3.14.16: Generators folder: /home/pieter/.conan2/p/b/ipopte17a619841258/b
ipopt/3.14.16: Generating aggregated env files
ipopt/3.14.16: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
ipopt/3.14.16: Calling build()
ipopt/3.14.16: RUN: "/home/pieter/.conan2/p/b/ipopte17a619841258/b/configure" --disable-shared --enable-static --prefix=/ '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--with-lapack= ' --with-mumps=yes
[...]
configure: Configuration of Ipopt successful

ipopt/3.14.16: RUN: make -j8
Making all in src
make[1]: Entering directory '/home/pieter/.conan2/p/b/ipopte17a619841258/b/src'
[...]
make[1]: Leaving directory '/home/pieter/.conan2/p/b/ipopte17a619841258/b'

ipopt/3.14.16: package(): Packaged 6 '.h' files
ipopt/3.14.16: package(): Packaged 79 '.hpp' files
ipopt/3.14.16: package(): Packaged 1 '.inc' file: IpReturnCodes.inc
ipopt/3.14.16: package(): Packaged 2 '.la' files: libipopt.la, libsipopt.la
ipopt/3.14.16: package(): Packaged 2 '.a' files: libipopt.a, libsipopt.a
ipopt/3.14.16: package(): Packaged 1 '.pc' file: ipopt.pc
ipopt/3.14.16: package(): Packaged 2 '.md' files: README.md, ChangeLog.md
ipopt/3.14.16: package(): Packaged 2 files: AUTHORS, LICENSE
ipopt/3.14.16: Created package revision 8d9eadcf5fd1383a17bc78f0650320c3
ipopt/3.14.16: Package 'aee99e0e9ca0db3bc3f87a71d61629d23c32e9b4' created
ipopt/3.14.16: Full package reference: ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c:aee99e0e9ca0db3bc3f87a71d61629d23c32e9b4#8d9eadcf5fd1383a17bc78f0650320c3
ipopt/3.14.16: Package folder /home/pieter/.conan2/p/b/ipopte17a619841258/p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: openblas/0.3.27
WARN: deprecated:     'env_info' used in: openblas/0.3.27

======== Launching test_package ========

======== Computing dependency graph ========
Graph root
    ipopt/3.14.16 (test package): /tmp/conan-recipes/recipes/ipopt/all/test_package/conanfile.py
Requirements
    coinmumps/3.0.7#53df8e3d1b90efec14bba86a1bf391c2 - Cache
    ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c - Cache
    openblas/0.3.27#c16c3dbdf45e4a4ff56f8ada8631a2f3 - Cache

======== Computing necessary packages ========
openblas/0.3.27: Main binary package 'bc6c7a69ae72a75e92e2dcbdc9e3560daef36419' missing
openblas/0.3.27: Checking 11 compatible configurations
openblas/0.3.27: Found compatible package 'eb6292a0e957646b54c2bfd72a2a6c8d0b7a1aac': compiler.cppstd=17
Requirements
    coinmumps/3.0.7#53df8e3d1b90efec14bba86a1bf391c2:da8cda98ce5c347f86e367506d27090df7150d5e#17436017e49b2b4c6eb2d8c6b264cc14 - Cache
    ipopt/3.14.16#01c815c005a134ac871a5d8b8bd8191c:aee99e0e9ca0db3bc3f87a71d61629d23c32e9b4#8d9eadcf5fd1383a17bc78f0650320c3 - Cache
    openblas/0.3.27#c16c3dbdf45e4a4ff56f8ada8631a2f3:eb6292a0e957646b54c2bfd72a2a6c8d0b7a1aac#6231bd84d94ddb4211bfaa4d6287e04c - Cache

======== Installing packages ========
openblas/0.3.27: Already installed! (1 of 3)
coinmumps/3.0.7: Already installed! (2 of 3)
ipopt/3.14.16: Already installed! (3 of 3)
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: openblas/0.3.27
WARN: deprecated:     'env_info' used in: openblas/0.3.27

======== Testing the package ========
Removing previously existing 'test_package' build folder: /tmp/conan-recipes/recipes/ipopt/all/test_package/build/gcc-11-x86_64-gnu17-release
ipopt/3.14.16 (test package): Test package build: build/gcc-11-x86_64-gnu17-release
ipopt/3.14.16 (test package): Test package build folder: /tmp/conan-recipes/recipes/ipopt/all/test_package/build/gcc-11-x86_64-gnu17-release
ipopt/3.14.16 (test package): Writing generators to /tmp/conan-recipes/recipes/ipopt/all/test_package/build/gcc-11-x86_64-gnu17-release/generators
ipopt/3.14.16 (test package): WARN: Using the new CMakeDeps generator, behind the 'tools.cmake.cmakedeps:new' gate conf. This conf will changenext release, breaking, so use it only for testing and dev
ipopt/3.14.16 (test package): Generator 'CMakeDeps' calling 'generate()'
ipopt/3.14.16 (test package): Generator 'CMakeToolchain' calling 'generate()'
ipopt/3.14.16 (test package): CMakeToolchain generated: conan_toolchain.cmake
ipopt/3.14.16 (test package): CMakeToolchain generated: /tmp/conan-recipes/recipes/ipopt/all/test_package/build/gcc-11-x86_64-gnu17-release/generators/CMakePresets.json
ipopt/3.14.16 (test package): CMakeToolchain generated: /tmp/conan-recipes/recipes/ipopt/all/test_package/CMakeUserPresets.json
ipopt/3.14.16 (test package): Generating aggregated env files
ipopt/3.14.16 (test package): Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']

======== Testing the package: Building ========
ipopt/3.14.16 (test package): Calling build()
ipopt/3.14.16 (test package): Running CMake.configure()
ipopt/3.14.16 (test package): RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/tmp/conan-recipes/recipes/ipopt/all/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/tmp/conan-recipes/recipes/ipopt/all/test_package"
-- Using Conan toolchain: /tmp/conan-recipes/recipes/ipopt/all/test_package/build/gcc-11-x86_64-gnu17-release/generators/conan_toolchain.cmake
-- Conan toolchain: Defining architecture flag: -m64
-- Conan toolchain: C++ Standard 17 with extensions ON
-- Conan toolchain: Including CMakeDeps generated conan_find_paths.cmake
-- The CXX compiler identification is GNU 11.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring Targets for ipopt/3.14.16
-- Configuring Targets for coinmumps/3.0.7
-- Configuring Targets for openblas/0.3.27
-- Conan: Target declared imported STATIC library 'OpenBLAS::pthread'
-- Conan: Target declared imported INTERFACE library 'OpenBLAS::OpenBLAS'
-- Conan: Target declared imported STATIC library 'coinmumps::coinmumps'
-- Conan: Target declared imported STATIC library 'Ipopt::Ipopt'
-- Configuring done (0.2s)
CMake Error at build/gcc-11-x86_64-gnu17-release/generators/coinmumps-Targets-release.cmake:20 (target_link_libraries):
  The link interface of target "coinmumps::coinmumps" contains:

    openblas::openblas

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  build/gcc-11-x86_64-gnu17-release/generators/coinmumpsTargets.cmake:6 (include)
  build/gcc-11-x86_64-gnu17-release/generators/coinmumps-config.cmake:6 (include)
  /home/pieter/.local/lib/python3.13/site-packages/cmake/data/share/cmake-3.28/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  build/gcc-11-x86_64-gnu17-release/generators/Ipopt-Targets-release.cmake:7 (find_dependency)
  build/gcc-11-x86_64-gnu17-release/generators/IpoptTargets.cmake:6 (include)
  build/gcc-11-x86_64-gnu17-release/generators/IpoptConfig.cmake:6 (include)
  CMakeLists.txt:4 (find_package)


CMake Error at build/gcc-11-x86_64-gnu17-release/generators/Ipopt-Targets-release.cmake:25 (target_link_libraries):
  The link interface of target "Ipopt::Ipopt" contains:

    openblas::openblas

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  build/gcc-11-x86_64-gnu17-release/generators/IpoptTargets.cmake:6 (include)
  build/gcc-11-x86_64-gnu17-release/generators/IpoptConfig.cmake:6 (include)
  CMakeLists.txt:4 (find_package)


-- Generating done (0.0s)
CMake Generate step failed.  Build files cannot be regenerated correctly.

ERROR: ipopt/3.14.16 (test package): Error in build() method, line 20
	cmake.configure()
	ConanException: Error 1 while executing
```

</details>

I'm using Conan 2.9.2 on Ubuntu 22.04 (x86_64), Python 3.10.14, CMake 3.28.4.

--- a/conan/tools/cmake/cmakedeps2/cmakedeps.py
+++ b/conan/tools/cmake/cmakedeps2/cmakedeps.py
@@ -191,9 +191,10 @@ class _PathGenerator:
                     build_dir = dep.package_folder
                 pkg_folder = build_dir.replace("\\", "/") if build_dir else None
                 if pkg_folder:
-                    config_file = ConfigTemplate2(self._cmakedeps, dep).filename
-                    if os.path.isfile(os.path.join(pkg_folder, config_file)):
-                        pkg_paths[pkg_name] = pkg_folder
+                    f = self._cmakedeps.get_cmake_filename(dep)
+                    for filename in (f"{f}-config.cmake", f"{f}Config.cmake"):
+                        if os.path.isfile(os.path.join(pkg_folder, filename)):
+                            pkg_paths[pkg_name] = pkg_folder
                 continue
 
             # If CMakeDeps generated, the folder is this one

--- a/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new_paths.py
+++ b/test/functional/toolchains/cmake/cmakedeps/test_cmakedeps_new_paths.py
@@ -47,10 +47,12 @@ def test_cmake_generated(client):
 
 
 @pytest.mark.tool("cmake")
-def test_cmake_in_package(client):
+@pytest.mark.parametrize("lowercase", [False, True])
+def test_cmake_in_package(client, lowercase):
     c = client
     # same, but in-package
-    dep = textwrap.dedent("""
+    f = "dep-config" if lowercase else "depConfig"
+    dep = textwrap.dedent(f"""
         import os
         from conan import ConanFile
         from conan.tools.files import save
@@ -60,7 +62,7 @@ def test_cmake_in_package(client):
 
             def package(self):
                 content = 'message(STATUS "Hello from dep dep-Config.cmake!!!!!")'
-                save(self, os.path.join(self.package_folder, "cmake", "dep-config.cmake"), content)
+                save(self, os.path.join(self.package_folder, "cmake", "{f}.cmake"), content)
             def package_info(self):
                 self.cpp_info.set_property("cmake_find_mode", "none")
                 self.cpp_info.builddirs = ["cmake"]

